doc/developer/design/20240411_graceful_reconfiguration_of_managed_clusters_v1.md

```diff
@@ -0,0 +1,331 @@
+to additionally check `check_reconfiguration_policy`, which sends
+`ScheduleDecisions` with the decision `FinalizeReconfigure`, when the reconfigure timeout deadline has passed.
+
+`check_reconfiguration_policy`, for V1, will trigger a `FinalizeReconfiguration` `ScheduingDecision` for clusters that
+have a `Reconfigure` `ClusterSchedule` with a timeout value greater than the current time. In V2, we will add logic to this to also check for parity/hydration. At this point, it may need to be backgrounded as it's likely a much more costly check.
+
+`Message::SchedulingDecisions` will be adjusted from 
+`SchedulingDecisions(Vec<(&'static str, Vec<(ClusterId, bool)>)>)` to be
+`SchedulingDecisions(Vec<SchedulingPolicyDecision>)`
+```rust
+enum SchedulingPolicyDecision {
+  Refresh(Vec<(ClusterId, bool)>),
+  FinalizeReconfigure(Vec<(ClusterId)>),
+}
+```
+
+The logic for `handle_schedule_decision` will be directed to the correct
+function based on the  variant of decision `handle_refresh_decision` or
+`handle_finalize_reconfigure_decision`.
+
+For each cluster `handle_finalize_reconfigure_decisions` will check the current
+catalog to avoid conflicts with catalog updates since the `FinalizeDecision`
+was sent. If there's a collision that should prevent the finalization, such as a
+new size adjustment, then break; Otherwise, remove the active replicas and move
+the reconfiguration replicas to be active. Finally, remove the `Reconfigure`
+schedule from the cluster.
+
+
+#### Cluster Sequencer
+The following roughly defines the new logic for the cluster sequencer when
+performing an alter cluster on a managed cluster.
+
+There are two states to look at
+1. the newly provided config
+2. the reconfigue config (what is in the catalog)
+
+First scenario, the cluster being altered has no `Reconfigure` `cluster_schedule`:
+If cleanup_timeout is None:
+ - use the old mechanism ( drop and replace )
+If the new alter statement provides a `cleanup_timeout`: 
+ - check for sources/sinks,
+ - update the catalog with the changes and add a `Reconfigure` cluster schedule
+ - deploy new reconfigure replicas
+
+Second scenario, the cluster being altered is undergoing a reconfigure:
+If cleanup_timeout is None,
+ - Use the old mechanism - Drop and replace including all reconfigure replicas and remove the schedule.
+If the new alter statement provides a `cleanup_timeout`:
+ - If the new config matches the current config
+   - bump the timeout
+ - If the new config doesn't match the current config
+   - check for sources/sinks
+   - update cluster catalog with new schedule and config
+   - drop reconfigure replicas
+   - launch new reconfigure replicas
+
+Note:
+An `ALTER CLUSTER` statement that is run twice (double shot), either needs to
+push out the timeout of an ongoing reconfiguration or needs to entirely replace
+the existing reconfigure schedule dropping and recreating any reconfigure
+replicas. I believe the former is the more useful behavior.
```

Comment:
I'd like to propose an alternative v1: just block the `ALTER CLUSTER` command until the cutover is complete. I [mentioned this briefly on the issue](https://github.com/MaterializeInc/materialize/issues/20010#issuecomment-2058302174):

> Should the ALTER CLUSTER statement block until the reconfiguration is complete? This is the "correct" behavior from the PostgreSQL perspective. For example, CREATE INDEX and ALTER TABLE both block until the potentially very slow operation is complete. This has the appealing property of allowing us to report failures to complete the operation back to the user. It also seems a good bit simpler to build, because we don't need to build any job scheduling infrastructure.

I think it'd be muuuuch simpler to start like this, since you basically wouldn't need any durable state to record that the alteration was in progress. Here's how it'd work:

1. `ALTER CLUSTER` command is planned normally, resulting in a call to `sequence_alter_cluster` on the coordinator main loop.
2. If `ALTER CLUSTER` command involves a reconfiguration, add the new replicas with the `-pending` suffix (and `pending = true` in the `mz_cluster_replicas` table) in a catalog transaction. In the same catalog transaction, ensure there are no existing pending replicas—if there are, send back an error, as another `ALTER CLUSTER` operation on the same cluster is already in progress. (In the future, we might prefer to have subsequent `ALTER CLUSTER` operations just _block_ until the earlier operation completes, but that's more work to rig up.)
3. Off of the coordinator main loop, wait out the `WAIT` policy.
4. Once the `WAIT` policy completes, issue another catalog transaction (on the coordinator main thread) that drops the old replicas as removes the `pending` bit from the new replicas.
5. Report success back to the user.

If the client goes away while the `ALTER` is running, we can have the coordinator abort the operation: just drop all the outstanding pending replicas. 

I don't think you'd need to interact with the cluster scheduler at all, other than to make sure that the schedule policy wasn't `ON REFRESH`. The interaction between `SCHEDULE = ON REFRESH` seems real tricky, and since `REFRESH EVERY` views are generally for cold/warm data rather than hot data, the lack of graceful cluster reconfiguration seems quite tolerable.

There's only one subtle bit: what if `envd` crashes in the middle of handling an `ALTER CLUSTER` command? E.g., say the `ALTER CLUSTER` is being executed during the maintenance window. I think the answer is pretty clean: just have `envd` drop all pending replicas when it boots. We _know_ that if we have pending replicas, the old `envd` 1) was in the middle of processing an `ALTER CLUSTER` command and 2) returned a "connection closed" error or somesuch that did not indicate success, and so we can simply drop the pending replicas and pretend the alteration never happend.

And, if blocking `ALTER CLUSTER` statements _does_ turn out to be a pain point, we can roll out support for `ALTER CLUSTER ... (BACKGROUND = TRUE)` in some future release of Materialize.

Reply:
I have some concerns about implementing non-blocking as follow up. As someone who regularly deals with cloud infra I find it much harder to work with things that must be handled over a single connection. Networks are flaky, laptops go to sleep, and connection timeouts threaten longer lived calls. 

This might work fine over PGWIRE where it's not abnormal to have long running queries (10s of minutes), but HTTP calls via the console may not be as resilient here. The connections themselves may be fine if they don't hit timeouts, but may find ourselves fighting with user interactions with long lived connections in the browser.